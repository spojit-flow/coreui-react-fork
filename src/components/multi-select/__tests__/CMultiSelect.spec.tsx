import * as React from 'react'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { CMultiSelect } from '../../../index'

const options = [
  {
    value: 0,
    text: 'Angular',
  },
  {
    value: 1,
    text: 'Bootstrap',
    selected: true,
  },
  {
    value: 2,
    text: 'React.js',
  },
  {
    value: 3,
    text: 'Vue.js',
    selected: true,
  },
  {
    label: 'backend',
    options: [
      {
        value: 4,
        text: 'Django',
      },
      {
        value: 5,
        text: 'Laravel',
      },
      {
        value: 6,
        text: 'Node.js',
        disabled: true,
        selected: true,
      },
    ],
  },
]

test('loads and displays CMultiSelect component', async () => {
  const { container } = render(<CMultiSelect options={[]} />)
  expect(container).toMatchSnapshot()
})

test('CMultiSelect customize', async () => {
  const { container } = render(
    <CMultiSelect
      options={options}
      className="bazinga"
      cleaner={true}
      multiple={true}
      optionsMaxHeight={150}
      optionsStyle="checkbox"
      placeholder="placeholder"
      search={true}
      searchNoResultsLabel="searchNoResultsLabel"
      selectAll={true}
      selectAllLabel="selectAllLabel"
      selectionType="tags"
      selectionTypeCounterText="selectionTypeCounterText"
      visible={true}
    />,
  )
  expect(container).toMatchSnapshot()
  expect(container.firstChild).toHaveAttribute('multiple', '')
  expect(container.firstChild).toHaveAttribute('style', 'display: none;')
  expect(container.firstChild).toHaveAttribute('tabindex', '-1')
  if (container.firstChild === null) {
    expect(true).toBe(false)
  } else {
    expect(container.firstChild.firstChild).toHaveAttribute('value', '1')
    expect(container.firstChild.firstChild).toHaveTextContent('Bootstrap')
    expect(container.firstChild.lastChild).toHaveAttribute('value', '6')
    expect(container.firstChild.lastChild).toHaveTextContent('Node.js')
  }
  expect(container.lastChild).toHaveClass('form-multi-select')
  expect(container.lastChild).toHaveClass('show')
  expect(container.lastChild).toHaveClass('bazinga')

  if (container.lastChild === null) {
    expect(true).toBe(false)
  } else {
    expect(container.lastChild.firstChild).toHaveClass('form-multi-select-toggler')
    if (container.lastChild.firstChild === null) {
      expect(true).toBe(false)
    } else {
      expect(container.lastChild.firstChild.firstChild).toHaveClass(
        'form-multi-select-selection-tags',
      )
      if (container.lastChild.firstChild.firstChild === null) {
        expect(true).toBe(false)
      } else {
        expect(container.lastChild.firstChild.firstChild.firstChild).toHaveClass(
          'form-multi-select-tag',
        )
        expect(container.lastChild.firstChild.firstChild.firstChild).toHaveTextContent('Bootstrap')
      }
    }
    expect(container.lastChild.lastChild).toHaveClass('form-multi-select-dropdown')
    if (container.lastChild.lastChild === null) {
      expect(true).toBe(false)
    } else {
      expect(container.lastChild.lastChild.firstChild).toHaveClass('form-multi-select-all')
      expect(container.lastChild.lastChild.firstChild).toHaveAttribute('type', 'button')
      expect(container.lastChild.lastChild.firstChild).toHaveTextContent('selectAllLabel')
      expect(container.lastChild.lastChild.lastChild).toHaveClass('form-multi-select-options')
      expect(container.lastChild.lastChild.lastChild).toHaveAttribute(
        'style',
        'max-height: 150px; overflow: scroll;',
      )
      if (container.lastChild.lastChild.lastChild === null) {
        expect(true).toBe(false)
      } else {
        expect(container.lastChild.lastChild.lastChild.firstChild).toHaveClass(
          'form-multi-select-option',
        )
        expect(container.lastChild.lastChild.lastChild.firstChild).toHaveClass(
          'form-multi-select-option-with-checkbox',
        )
        expect(container.lastChild.lastChild.lastChild.firstChild).toHaveTextContent('Angular')
        expect(container.lastChild.lastChild.lastChild.lastChild).toHaveClass(
          'form-multi-select-option',
        )
        expect(container.lastChild.lastChild.lastChild.lastChild).toHaveClass(
          'form-multi-select-option-with-checkbox',
        )
        expect(container.lastChild.lastChild.lastChild.lastChild).toHaveClass('form-multi-selected')
        expect(container.lastChild.lastChild.lastChild.lastChild).toHaveClass('disabled')
        expect(container.lastChild.lastChild.lastChild.lastChild).toHaveTextContent('Node.js')
      }
    }
  }
})

test('CMultiSelect fire events', async () => {
  const event = jest.fn()
  expect(event).toHaveBeenCalledTimes(0)
  render(
    <CMultiSelect
      options={options}
      className="bazinga"
      cleaner={true}
      multiple={true}
      optionsMaxHeight={150}
      optionsStyle="checkbox"
      placeholder="placeholder"
      search={true}
      searchNoResultsLabel="searchNoResultsLabel"
      selectAll={true}
      selectAllLabel="selectAllLabel"
      selectionType="tags"
      selectionTypeCounterText="selectionTypeCounterText"
      visible={true}
      onChange={event}
    />,
  )
  expect(event).toHaveBeenCalledTimes(1)
  const option = document.querySelector('.form-multi-select-option')
  if (option !== null) {
    fireEvent.click(option)
  }
  expect(event).toHaveBeenCalledTimes(2)
  const removeButton = document.querySelector('.form-multi-select-tag-delete')
  if (removeButton !== null) {
    fireEvent.click(removeButton)
  }
  expect(event).toHaveBeenCalledTimes(3)
})
