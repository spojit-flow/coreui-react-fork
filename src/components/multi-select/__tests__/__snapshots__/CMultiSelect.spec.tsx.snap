// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CMultiSelect customize 1`] = `
<div>
  <select
    class="multi-select-new"
    multiple=""
    style="display: none;"
    tabindex="-1"
  >
    <option
      value="1"
    >
      Bootstrap
    </option>
    <option
      value="3"
    >
      Vue.js
    </option>
    <option
      disabled=""
      value="6"
    >
      Node.js
    </option>
  </select>
  <div
    aria-expanded="true"
    class="form-multi-select show bazinga"
  >
    <div
      class="form-multi-select-toggler"
    >
      <span
        class="form-multi-select-selection form-multi-select-selection-tags"
      >
        <span
          class="form-multi-select-tag"
        >
          Bootstrap
          <button
            aria-label="Close"
            class="form-multi-select-tag-delete"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </span>
        <span
          class="form-multi-select-tag"
        >
          Vue.js
          <button
            aria-label="Close"
            class="form-multi-select-tag-delete"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
          </button>
        </span>
        <span
          class="form-multi-select-tag"
        >
          Node.js
        </span>
        <input
          autocomplete="off"
          class="form-multi-select-search"
          size="2"
          type="text"
        />
      </span>
      <div
        class="form-multi-select-buttons"
      >
        <button
          class="form-multi-select-cleaner"
          type="button"
        />
        <button
          class="form-multi-select-indicator"
          type="button"
        />
      </div>
    </div>
    <div
      class="form-multi-select-dropdown"
      role="menu"
      style="position: absolute; left: 0px; top: 0px; margin: 0px;"
    >
      <button
        class="form-multi-select-all"
        type="button"
      >
        selectAllLabel
      </button>
      <div
        class="form-multi-select-options"
        style="max-height: 150px; overflow: scroll;"
      >
        <div
          class="form-multi-select-option form-multi-select-option-with-checkbox"
          tabindex="0"
        >
          Angular
        </div>
        <div
          class="form-multi-select-option form-multi-select-option-with-checkbox form-multi-selected"
          tabindex="0"
        >
          Bootstrap
        </div>
        <div
          class="form-multi-select-option form-multi-select-option-with-checkbox"
          tabindex="0"
        >
          React.js
        </div>
        <div
          class="form-multi-select-option form-multi-select-option-with-checkbox form-multi-selected"
          tabindex="0"
        >
          Vue.js
        </div>
        <div
          class="form-multi-select-optgroup-label"
        >
          backend
        </div>
        <div
          class="form-multi-select-option form-multi-select-option-with-checkbox"
          tabindex="0"
        >
          Django
        </div>
        <div
          class="form-multi-select-option form-multi-select-option-with-checkbox"
          tabindex="0"
        >
          Laravel
        </div>
        <div
          class="form-multi-select-option form-multi-select-option-with-checkbox form-multi-selected disabled"
          tabindex="0"
        >
          Node.js
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`loads and displays CMultiSelect component 1`] = `
<div>
  <select
    class="multi-select-new"
    multiple=""
    style="display: none;"
    tabindex="-1"
  />
  <div
    aria-expanded="false"
    class="form-multi-select"
  >
    <div
      class="form-multi-select-toggler"
    >
      <span
        class="form-multi-select-selection form-multi-select-selection-tags"
      >
        <input
          autocomplete="off"
          class="form-multi-select-search"
          placeholder="Select..."
          type="text"
        />
      </span>
      <div
        class="form-multi-select-buttons"
      >
        <button
          class="form-multi-select-indicator"
          type="button"
        />
      </div>
    </div>
    <div
      class="form-multi-select-dropdown"
      role="menu"
    >
      <button
        class="form-multi-select-all"
        type="button"
      >
        Select all options
      </button>
      <div
        class="form-multi-select-options"
      >
        <div
          class="form-multi-select-options-empty"
        >
          No results found
        </div>
      </div>
    </div>
  </div>
</div>
`;
